sky/engine/core/script/dart_controller.cc

```diff
@@ -78,6 +63,75 @@ DartController::~DartController() {
   }
 }
 
+static Dart_Handle CreateMainArguments(const std::vector<std::string>& args) {
+  Dart_Handle dart_arguments = Dart_NewList(args.size());
+  if (Dart_IsError(dart_arguments)) {
+    return dart_arguments;
+  }
+  for (size_t i = 0; i < args.size(); i++) {
+    Dart_Handle argument_value = ToDart(args[i]);
+    if (Dart_IsError(argument_value)) {
+      return argument_value;
+    }
+    Dart_Handle result = Dart_ListSetAt(dart_arguments, i, argument_value);
+    if (Dart_IsError(result)) {
+      return result;
+    }
+  }
+  return dart_arguments;
+}
+
+bool DartController::SendStartMessage(Dart_Handle root_library,
+                                      const std::vector<std::string>& args) {
+  {
+    // Temporarily exit the isolate while we make it runnable.
+    Dart_Isolate isolate = dart_state()->isolate();
+    DCHECK(Dart_CurrentIsolate() == isolate);
+    Dart_ExitIsolate();
+    CHECK(Dart_IsolateMakeRunnable(isolate));
+    Dart_EnterIsolate(isolate);
+  }
+
+  // In order to support pausing the isolate at start, we indirectly invoke
+  // main by sending a message to the isolate.
+  // Grab the 'dart:ui' library.
+  Dart_Handle ui_library = Dart_LookupLibrary(ToDart("dart:ui"));
+  DART_CHECK_VALID(ui_library);
+
+  // Grab the 'dart:isolate' library.
+  Dart_Handle isolate_lib = Dart_LookupLibrary(ToDart("dart:isolate"));
+  DART_CHECK_VALID(isolate_lib);
+
+  // Import the root library into the 'dart:ui' library so that we can
+  // reach main.
+  Dart_LibraryImportLibrary(ui_library, root_library, Dart_Null());
+
+  // Get the closure of main().
+  Dart_Handle main_closure = Dart_Invoke(ui_library,
+                                         ToDart("_getMainClosure"),
+                                         0,
+                                         NULL);
+  DART_CHECK_VALID(main_closure);
+  // Create the arguments list.
+  Dart_Handle arguments = CreateMainArguments(args);
+  // Send the start message containing the entry point + arguments by calling
+  // _startMainIsolate in dart:isolate.
+  const intptr_t kNumIsolateArgs = 2;
+  Dart_Handle isolate_args[kNumIsolateArgs];
+  isolate_args[0] = main_closure;
+  isolate_args[1] = arguments;
+  Dart_Handle result = Dart_Invoke(isolate_lib,
+                                   ToDart("_startMainIsolate"),
+                                   kNumIsolateArgs,
+                                   isolate_args);
+  return LogIfError(result);
+}
+
+bool DartController::SendStartMessage(Dart_Handle root_library) {
+  std::vector<std::string> args;
+  return SendStartMessage(root_library, args);
```

Comment:
Can we just remove all this machinery for passing arguments to main?  We don't pass arguments to main.